### PR TITLE
feat(experiments): Increase max number of variants

### PR DESCRIPTION
**Pull request by neilkakkar**
## Problem

Allow for more variants in experiments

TODO: Sort out frontend, add flag, add colours

Things do look a bit more busy on the frontend

image

(cc: @corywatilo do you have ideas on how to make this look nicer? - context: Don't want to polish everything up, but just want to get this out for a customer who really wants this)

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

**Comment by neilkakkar** [comment]
image

File: ee/clickhouse/queries/experiments/funnel_experiment_result.py [hunk 1]
```diff
@@ -144,8 +144,8 @@ def calculate_results(
         if not control_variant:
             raise ValidationError("No control variant data found", code="no_data")
 
-        if len(test_variants) > 3:
-            raise ValidationError("Can't calculate A/B test results for more than 4 variants", code="too_much_data")
+        if len(test_variants) >= 10:
```
Comment (liyiy) [inline ee/clickhouse/queries/experiments/funnel_experiment_result.py:147]: How much slower will this cause things to be?

**Comment by liyiy** [comment]
If there's that many variants to display it might make sense to display them in either a table or chart ! LMK if you'd want some help with that.

**Comment by corywatilo** [comment]
Two things that would help:

1. Use the same table grid from the secondary metrics as the _probability_ view. That way, you don't have to repeat the headings.
2. In the cells about _probability_, don't include the text "Probability that this variant is the best:" - that only needs to be there for the legend in the left-most cell, so only the value lives in those columns

Those two tweaks and it will feel 💸

File: ee/clickhouse/queries/experiments/funnel_experiment_result.py (hunk 1 above)
Comment (neilkakkar) [reply ee/clickhouse/queries/experiments/funnel_experiment_result.py:147]: Not as bad as I was expecting, since we've switched how we compute results to be a monte-carlo simulation vs a closed-form solution I was trying before! (this happened during initial implementation, I just forgot that it means the limit of 3 isn't necessary anymore)

TL;DR -> It's almost no difference to before, as the limiting factor here is getting the insight results, not this computation.

**Comment by neilkakkar** [comment]
Thanks!

image

File: frontend/src/scenes/experiments/Experiment.tsx [hunk 2]
```diff
@@ -817,8 +829,79 @@ export function Experiment(): JSX.Element {
                                             <ExperimentImplementationDetails experiment={experimentData} />
                                         </Col>
                                     )}
-                                    {(experimentResults || experimentData.secondary_metrics?.length > 0) && (
-                                        <Col className="secondary-progress" span={experimentData?.start_date ? 12 : 24}>
+                                    {experimentResults && (
+                                        <Col span={12} className="mt-4">
```
Comment (neilkakkar) [inline frontend/src/scenes/experiments/Experiment.tsx:833]: this bit is just moved, no changes

File: frontend/src/scenes/experiments/Experiment.tsx [hunk 3]
```diff
@@ -1024,7 +1027,73 @@ export function Experiment(): JSX.Element {
                     </Row>
                     <div className="experiment-result">
                         {experimentResults ? (
-                            <>
+                            experimentData?.parameters?.feature_flag_variants?.length || 0 > 4 ? (
```
Comment (neilkakkar) [inline frontend/src/scenes/experiments/Experiment.tsx:1030]: This is the main conditional for the experiment results view:

When less than 4 total variants, we stick to the old view

When mooar, we switch to table view

Comment (liyiy) [reply frontend/src/scenes/experiments/Experiment.tsx:1030]: A vertical table might be a lot easier to look at rather than a horizontal one ! 😃

Comment (neilkakkar) [reply frontend/src/scenes/experiments/Experiment.tsx:1030]: Oh, interesting, that might be true.
 
 I'd rather not change again, unless you think this is blocking? 😅 .

Comment (liyiy) [reply frontend/src/scenes/experiments/Experiment.tsx:1030]: It's probably true lol but if it's emergency though, non-blocking

**Review by liyiy — APPROVED** [review]

File: frontend/src/scenes/experiments/Experiment.tsx (hunk 3 above)
Comment (liyiy) [reply frontend/src/scenes/experiments/Experiment.tsx:1030]: Like using a lemon table or one of our existing table views

Comment (neilkakkar) [reply frontend/src/scenes/experiments/Experiment.tsx:1030]: oh, that makes sense 😅 - should've started with that
